src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm

```diff
@@ -112,9 +115,91 @@
+    const chip::ByteSpan & csrNonce, const chip::ByteSpan & csrElementsSignature, const chip::ByteSpan & attestationChallenge,
+    const chip::ByteSpan & DAC, const chip::ByteSpan & PAI,
+    chip::Callback::Callback<chip::Controller::OnNOCChainGeneration> * onCompletion)
+{
+    mOnNOCCompletionCallback = onCompletion;
+
+    TLVReader reader;
+    reader.Init(csrElements);
+
+    if (reader.GetType() == kTLVType_NotSpecified) {
+        ReturnErrorOnFailure(reader.Next());
+    }
+
+    VerifyOrReturnError(reader.GetType() == kTLVType_Structure, CHIP_ERROR_WRONG_TLV_TYPE);
+    VerifyOrReturnError(reader.GetTag() == AnonymousTag(), CHIP_ERROR_UNEXPECTED_TLV_ELEMENT);
+
+    TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+    ReturnErrorOnFailure(reader.Next(kTLVType_ByteString, TLV::ContextTag(1)));
+
+    chip::ByteSpan csr(reader.GetReadPoint(), reader.GetLength());
+    reader.ExitContainer(containerType);
+
+    CSRInfo * csrInfo = [[CSRInfo alloc] initWithNonce:AsData(csrNonce)
+                                              elements:AsData(csrElements)
+                                     elementsSignature:AsData(csrElementsSignature)
+                                                   csr:AsData(csr)];
+
+    chip::ByteSpan certificationDeclarationSpan;
+    chip::ByteSpan attestationNonceSpan;
+    uint32_t timestampDeconstructed;
+    chip::ByteSpan firmwareInfoSpan;
+    chip::Credentials::DeviceAttestationVendorReservedDeconstructor vendorReserved;
+
+    chip::Optional<chip::Controller::CommissioningParameters> commissioningParameters
+        = mCppCommissioner->GetCommissioningParameters();
+    VerifyOrReturnError(commissioningParameters.HasValue(), CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(
+        chip::Credentials::DeconstructAttestationElements(commissioningParameters.Value().GetAttestationElements().Value(),
+            certificationDeclarationSpan, attestationNonceSpan, timestampDeconstructed, firmwareInfoSpan, vendorReserved));
+
+    AttestationInfo * attestationInfo =
+        [[AttestationInfo alloc] initWithChallenge:AsData(attestationChallenge)
+                                             nonce:AsData(commissioningParameters.Value().GetAttestationNonce().Value())
+                                          elements:AsData(commissioningParameters.Value().GetAttestationElements().Value())
+                                 elementsSignature:AsData(commissioningParameters.Value().GetAttestationSignature().Value())
+                                               dac:AsData(DAC)
+                                               pai:AsData(PAI)
+                          certificationDeclaration:AsData(certificationDeclarationSpan)
+                                      firmwareInfo:AsData(firmwareInfoSpan)];
+
+    [mNocChainIssuer onNOCChainGenerationNeeded:csrInfo attestationInfo:attestationInfo];
```

Comment:
This is calling into the issuer while on the Matter queue, no?

That's extremely dangerous: any call back into an API that does a dispatch_sync to the Matter queue (of which there are many) will deadlock.

The issuer should come with a queue that its `onNOCChainGenerationNeeded` will be called on, so we are not calling out into arbitrary code while on the Matter queue.

Reply:
Updated, dispatching calls to this on the mNocChainIssuerQueue.